### PR TITLE
[AMCL] reduce the amount of memory used

### PR DESCRIPTION
**Pull request by facontidavide**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | none |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Simulation |
| Does this PR contain AI generated software? | No |

---

## Description of contribution in a few bullet points

This is a memory optimization that should not affect realtime perfomance, but will reduce considerably the amount of memory used by AMCL.

In our large map (200 Mb) this is the memory usage **before** the change:

And **after** the changes:

The main changes are:

1. Changes in `map_cell_t`: reduction from 12 bytes to 5 bytes
2. Add a new parameter to reduce the memory used by `free_space_indices` using 1/4th of the points (equally spaced)

## Description of documentation updates required from your changes

Describe the new parameter "freespace_downsampling"

---

## Future work that may be required in bullet points

To be discussed

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Comment by doisyg** [comment]
Noting that it correspond to a new (smaller) maximum side size of 3.2 km for AMCL maps at a standard resolution of 0.05m/px. I think that it is fair and that we don't have any use cases of nav2 with maps that big. 
I expect other part of the system to fail before a map side .
wdyt @SteveMacenski and @tonynajjar ?

**Comment by facontidavide** [comment]
UPDATE: sorry I don't understand why this PR include files outside **nav2_amcl**. I will clean it up

File: nav2_amcl/src/amcl_node.cpp [hunk 1]
```diff
@@ -1106,6 +1111,7 @@ AmclNode::initParameters()
   get_parameter("always_reset_initial_pose", always_reset_initial_pose_);
   get_parameter("scan_topic", scan_topic_);
   get_parameter("map_topic", map_topic_);
+  get_parameter("freespace_downsampling", freespace_downsampling_);
```
Comment (SteveMacenski) [inline nav2_amcl/src/amcl_node.cpp:1114]: Perhaps set this as a configuratable downsampling step? i.e. 2,3,4,5,... might be nice to experiment with yourself.

**Review by SteveMacenski — APPROVED** [review]
Otherwise LGTM. AMCL changes usually scare me since its been untouched for so long, but this is minor and I don't foresee any issues. 

`double->float` there's a reason MPPI/Smac both use floats for everything, I found massive improvements in performance on it alone, but I haven't had the cycles to try to apply that to the full stack to see how much more it could be reduced (though, those are 2 of the heaviest operations outside of the costmap).

Party on! What's next for memory consumption? :sunglasses:

**Comment by SteveMacenski** [comment]
`freespace_downsampling_` needs to be a class member, that should get CI unblocked

File: nav2_amcl/src/amcl_node.cpp (hunk 1 above)
Comment (facontidavide) [reply nav2_amcl/src/amcl_node.cpp:1114]: We have a comically large map and 1/th is already enough to make us happy

**Comment by SteveMacenski** [comment]
Still a draft, should I merge this in?

**Comment by facontidavide** [comment]
if it is good for you, it is good for me. I will prepare tomorrow a PR for the documentation https://docs.nav2.org/configuration/packages/configuring-amcl.html?highlight=amcl

**Comment by facontidavide** [comment]
> Party on! What's next for memory consumption? 😎

:smile: sparse costmaps ??? Not yet... not yet...

**Comment by SteveMacenski** [comment]
Please do update the docs for this param. You can put it into the Iron->Jazzy file since Nav2 hasn't forked for Jazzy yet (still working on it..)

> 😄 sparse costmaps ??? Not yet... not yet...

Well before you go nutty there, lets discuss that. I have a broader plan for costmaps in the medium term future :wink:

**Comment by facontidavide** [comment]
> Well before you go nutty there, lets discuss that. I have a broader plan for costmaps in the medium term future
Keep me in the loop!

**Pull request merged**